enterprise/server/hit_tracker_client/hit_tracker_client_test.go

```diff
@@ -1,15 +1,195 @@
+	require.Equal(t, int64(2000), hitTrackerService.bytesDownloaded[interfaces.AuthAnonymousUser].Load())
+}
+
+func TestCASHitTracker_NoDeduplication(t *testing.T) {
+	_, hitTrackerFactory, hitTrackerService, fakeClock := setup(t)
+	ctx := context.Background()
+	hitTracker := hitTrackerFactory.NewCASHitTracker(ctx, &repb.RequestMetadata{})
+	hitTracker.TrackDownload(aDigest).CloseWithBytesTransferred(1000, 2000, repb.Compressor_IDENTITY, "test")
+	hitTracker.TrackDownload(aDigest).CloseWithBytesTransferred(1000, 2000, repb.Compressor_IDENTITY, "test")
+
+	for i := 0; i < 10; i++ {
+		fakeClock.Advance(time.Second)
+		time.Sleep(5 * time.Millisecond)
+	}
+
+	expectToEqual(t, int64(2), hitTrackerService.downloads[interfaces.AuthAnonymousUser], "Expected 2 cache hits")
+	require.Equal(t, int64(4000), hitTrackerService.bytesDownloaded[interfaces.AuthAnonymousUser].Load())
+}
+
+func TestCASHitTracker_DropsUpdates(t *testing.T) {
+	authenticator, hitTrackerFactory, hitTrackerService, fakeClock := setup(t)
+	ctx := context.Background()
+	hitTracker := hitTrackerFactory.NewCASHitTracker(ctx, &repb.RequestMetadata{})
+	for i := 0; i < 100; i++ {
+		hitTracker.TrackDownload(aDigest).CloseWithBytesTransferred(1, 2, repb.Compressor_IDENTITY, "test")
+		hitTracker.TrackDownload(bDigest).CloseWithBytesTransferred(10, 20, repb.Compressor_IDENTITY, "test")
+		hitTracker.TrackDownload(cDigest).CloseWithBytesTransferred(100, 200, repb.Compressor_IDENTITY, "test")
+		hitTracker.TrackDownload(dDigest).CloseWithBytesTransferred(1_000, 2_000, repb.Compressor_IDENTITY, "test")
+		hitTracker.TrackDownload(eDigest).CloseWithBytesTransferred(10_000, 20_000, repb.Compressor_IDENTITY, "test")
+	}
+
+	group1Ctx := authenticator.AuthContextFromAPIKey(context.Background(), user1)
+	hitTracker = hitTrackerFactory.NewCASHitTracker(group1Ctx, &repb.RequestMetadata{})
+	hitTracker.TrackDownload(fDigest).CloseWithBytesTransferred(1_000_000, 2_000_000, repb.Compressor_IDENTITY, "test")
+
+	for i := 0; i < 10; i++ {
+		fakeClock.Advance(time.Second)
+		time.Sleep(5 * time.Millisecond)
+	}
+
+	// These updates should come in in-order, so they should be:
+	// ANON: a, b, c, d, e, a, b, c, d, e
+	expectToEqual(t, int64(10), hitTrackerService.downloads[interfaces.AuthAnonymousUser], "Expected 10 cache hits for ANON")
+	require.Equal(t, int64(44_444), hitTrackerService.bytesDownloaded[interfaces.AuthAnonymousUser].Load())
+
+	// Even though group 1's update came at the end, it should still be sent.
+	expectToEqual(t, int64(1), hitTrackerService.downloads[group1], "Expected 1 cache hits for group 1")
+	require.Equal(t, int64(2_000_000), hitTrackerService.bytesDownloaded[group1].Load())
+}
+
+func expectToEqual(t *testing.T, expected int64, actual *atomic.Int64, message string) {
+	backoff := time.Millisecond
+	for i := 0; i < 10; i++ {
+		if expected == actual.Load() {
+			return
+		}
+		time.Sleep(backoff)
+		backoff = backoff * 2
+	}
+	require.Fail(t, message)
```

Comment:
under load, this might be flaky since the waiting period is only 2s total. Could crank up 10 => 16 or so, or just remove the upperbound and let the test time out.

Reply:
Done